jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/ThreadIdPool.java

```diff
@@ -0,0 +1,244 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util.thread;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.eclipse.jetty.util.MemoryUtils;
+import org.eclipse.jetty.util.ProcessorUtils;
+import org.eclipse.jetty.util.component.Dumpable;
+import org.eclipse.jetty.util.component.DumpableCollection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A fixed sized pool of items that uses ThreadId to avoid contention.
+ * This class can be used, instead of a {@link ThreadLocal}, when pooling items
+ * that are expensive to create, but only used briefly in the scope of a single thread.
+ * It is safe to use with {@link org.eclipse.jetty.util.VirtualThreads}, as unlike a {@link ThreadLocal} pool,
+ * the number of items is limited.
+ * <p>This is a light-weight version of {@link org.eclipse.jetty.util.ConcurrentPool} that is best used
+ * when items do not reserve an index in the pool even when acquired.
+ * @see org.eclipse.jetty.util.ConcurrentPool
+ */
+public class ThreadIdPool<E> implements Dumpable
+{
+    private static final Logger LOG = LoggerFactory.getLogger(ThreadIdPool.class);
+
+    // How far the entries in the AtomicReferenceArray are spread apart to avoid false sharing;
+    // use getIntegersPerCacheLine() instead of getLongsPerCacheLine() b/c references could be compressed.
+    private static final int SPREAD_FACTOR = MemoryUtils.getIntegersPerCacheLine();
+
+    private final AtomicReferenceArray<E> _items;
+
+    public ThreadIdPool()
+    {
+        this(-1);
+    }
+
+    public ThreadIdPool(int capacity)
+    {
+        _items = new AtomicReferenceArray<>((calcCapacity(capacity) + 1) * SPREAD_FACTOR);
+        if (LOG.isDebugEnabled())
+            LOG.debug("{}", this);
+    }
+
+    private static int calcCapacity(int capacity)
+    {
+        if (capacity >= 0)
+            return capacity;
+        return 2 * ProcessorUtils.availableProcessors();
+    }
+
+    /**
+     * @return the maximum number of items
+     */
+    public int capacity()
+    {
+        return (_items.length() - 1) / SPREAD_FACTOR;
```

Comment:
If we normalize the `capacity` to a power of 2, all modulo operations will become a shift.
However, I don't think it's extremely necessary.

But, given how often we reference this field, perhaps it'll be good to cache it in a final field, rather than computing it every time.